src/Features/LanguageServer/Protocol/Handler/References/FindAllReferencesHandler.cs

```diff
@@ -0,0 +1,118 @@
+﻿// Copyright (c) Microsoft.  All Rights Reserved.  Licensed under the Apache License, Version 2.0.  See License.txt in the project root for license information.
+
+using System.Collections.Generic;
+using System.Composition;
+using System.Linq;
+using System.Threading;
+using System.Threading.Tasks;
+using Microsoft.CodeAnalysis.Classification;
+using Microsoft.CodeAnalysis.Editor.FindUsages;
+using Microsoft.CodeAnalysis.Editor.Shared.Extensions;
+using Microsoft.CodeAnalysis.FindUsages;
+using Microsoft.CodeAnalysis.PooledObjects;
+using Microsoft.VisualStudio.Text.Adornments;
+using LSP = Microsoft.VisualStudio.LanguageServer.Protocol;
+
+namespace Microsoft.CodeAnalysis.LanguageServer.Handler
+{
+    [Shared]
+    [ExportLspMethod(LSP.Methods.TextDocumentReferencesName)]
+    internal class FindAllReferencesHandler : IRequestHandler<LSP.ReferenceParams, object[]>
+    {
+        public async Task<object[]> HandleRequestAsync(Solution solution, LSP.ReferenceParams request,
+            LSP.ClientCapabilities clientCapabilities, CancellationToken cancellationToken)
+        {
+            var locations = ArrayBuilder<LSP.Location>.GetInstance();
+
+            var document = solution.GetDocumentFromURI(request.TextDocument.Uri);
+            if (document == null)
+            {
+                return locations.ToArrayAndFree();
+            }
+
+            var findUsagesService = document.Project.LanguageServices.GetService<IFindUsagesService>();
+            var position = await document.GetPositionFromLinePositionAsync(ProtocolConversions.PositionToLinePosition(request.Position), cancellationToken).ConfigureAwait(false);
+
+            var context = new SimpleFindUsagesContext(cancellationToken);
+
+            await findUsagesService.FindReferencesAsync(document, position, context).ConfigureAwait(false);
+
+            if (clientCapabilities?.HasVisualStudioLspCapability() == true)
```

Comment:
I think we need to talk with live share team whether this is the way we want for client capability. I think we would want feature specific capability at the end. work item?

Reply:
https://github.com/dotnet/roslyn/projects/45#card-20876090